// ngf-server/controllers/exportToDB.js
var events = require('../database/events'),
users = require('../database/users')

exports.exportData = function(req, res){
    console.log('connected')
    var sql = require('../../mssql')
      var config = {
          user: 'sa',
          password: '123456789',
          server: 'localhost\\MSSQLSERVER01',
          database: 'EventsDatabase',
          port: 1433
      }
      sql.connect(config, function (err) {

        if (err) console.log(err)

        var request = new sql.Request()

        console.log('connected')

        users.forEach(user => {
            request.batch(`INSERT INTO Users(id, Username, Password, FirstName, LastName) VALUES
            (${user.id},'${user.userName}', '${user.password}', '${user.firstName}', '${user.lastName}')`)
        })

        events.forEach(event => {
            if (event.location) {
                request.batch(`INSERT INTO Locations(Address, City, Country) VALUES ('${event.location.address}','${event.location.city}','${event.location.country}')`)
                request.batch(`INSERT INTO Events (Id, Name, Date, Time, Price, ImageUrl, LocationId, OnlineUrl) SELECT
                ${event.id}, '${event.name}','${event.date}', '${event.time}', ${event.price}, '${event.imageUrl}', id , '${event.onlineUrl}' FROM Locations WHERE Address = '${event.location.address}'
                AND City = '${event.location.city}' AND Country = '${event.location.country}'`)
            }
            else if(event.location == null){
                request.batch(`INSERT INTO Events (Id, Name, Date, Time, Price, ImageUrl, LocationId, OnlineUrl) VALUES
                (${event.id},'${event.name}','${event.date}', '${event.time}', ${event.price}, '${event.imageUrl}', NULL, '${event.onlineUrl}')`)
            }
            setTimeout(function(){event.sessions.forEach(session =>{
              request.batch(`INSERT INTO Sessions (Id ,Name, Presenter, Duration, Level, Abstract, EventId) VALUES
              (${session.id}, '${session.name}', '${session.presenter}', ${session.duration}, '${session.level}', '${session.abstract}', ${event.id})`)

              setTimeout(function(){session.voters.forEach(voter =>{
                request.batch(`INSERT INTO Voters(UserId, SessionId) SELECT
                Id, ${session.id}  FROM Users WHERE Username = '${voter}'`)
              })},3000)

          })          
          }, 3000)

      })
  })
  }
